library/Freckle/App/Kafka/Consumer.hs

```diff
@@ -164,31 +164,39 @@ runConsumer
      , HasCallStack
      )
   => Timeout
+  -> Int
   -> (a -> m ())
   -> m ()
-runConsumer pollTimeout onMessage =
+runConsumer pollTimeout batchSize onMessage =
   withTraceIdContext $ immortalCreate onFinish $ do
     consumer <- view kafkaConsumerL
 
     flip catches handlers $ inSpan "kafka.consumer" consumerSpanArguments $ do
-      mRecord <- fromKafkaError =<< pollMessage consumer kTimeout
-
-      for_ mRecord $ \r -> do
-        for_ (crValue r) $ \bs -> do
-          a <-
-            inSpan "kafka.consumer.message.decode" defaultSpanArguments $
-              either (throwM . KafkaMessageDecodeError bs) pure $
-                eitherDecodeStrict bs
-          inSpan "kafka.consumer.message.handle" defaultSpanArguments $ onMessage a
-
-        inSpan "kafka.consumer.message.commit" defaultSpanArguments $ do
-          -- Store the offset of this record, then do a best-offort commit to
-          -- the broker. If this fails and we crash or shutdown, the commit in
-          -- the onFinish handler will pick it up.
-          logExMay "Unable to store offset" $ storeOffsetMessage consumer r
-          void $ commitAllOffsets OffsetCommitAsync consumer
+      mRecords <-
+        traverse fromKafkaError
+          =<< pollMessageBatch consumer kTimeout kBatchSize
+
+      for_ mRecords $ \mRecord -> do
+        for_ mRecord $ \r -> do
+          for_ (crValue r) $ \bs -> do
+            a <-
+              inSpan "kafka.consumer.message.decode" defaultSpanArguments $
+                either (throwM . KafkaMessageDecodeError bs) pure $
+                  eitherDecodeStrict bs
+            inSpan "kafka.consumer.message.handle" defaultSpanArguments $ onMessage a
+            logExMay
+              "Unable to store offset"
+              (silenceNoOffsetError <$> storeOffsetMessage consumer r)
+
+      inSpan "kafka.consumer.message.commit" defaultSpanArguments $
+        -- Do a best-effort commit to the broker. If this fails and
+        -- we crash or shutdown, the commit in the onFinish handler
+        -- will pick it up.
+        void $
+          commitAllOffsets OffsetCommitAsync consumer
```

Comment:
I believe the `_NO_OFFSETS` error comes from here (when you commit with nothing stored), so you'd want to move your silence function, right?

Reply:
I'm seeing the `Unable to commit offsets` message which I think comes from the `logExMay "Unable to store offset"` when we `storeOffsetMessage`

Reply:
Sorry, I'm confused... you're saying you see `"Unable to commit offsets"` but you claim it comes from where we log `"Unable to store offset"`? That doesn't seem right, those are different messages.

Log message have source location though, so we shouldn't have to guess.

Reply:
Oh the looping is from `immortal`- we call `onFinish` after every message

Reply:
Wait- that means we have a synchronous offset commit on every message. Is that what we really want?

Reply:
Opened up https://github.com/freckle/freckle-app/pull/152 to explore fixing this